hilti/include/rt/types/set.h

```diff
@@ -63,36 +112,75 @@ class Set : protected std::set<T>, public hilti::rt::detail::iterator::Controlle
     Set& operator=(const Set&) = default;
     Set& operator=(Set&&) noexcept = default;
 
-    /** Returns true if a specific element is part of the set. */
-    bool contains(const T& t) { return this->find(t) != this->end(); }
-
-    // Methods of `std::set`.
-    using V::begin;
-    using V::cbegin;
-    using V::cend;
-    using V::end;
+    /** Checsk whether an element is in the set.
+     *
+     * @param `t` the element to check for
+     * @return `true` if the element is part of the set. */
+    bool contains(const T& t) { return this->count(t); }
+
+    auto begin() const { return iterator(static_cast<const V&>(*this).begin(), empty() ? nullptr : _control); }
```

Comment:
Not sure I fully follow here why we're not setting the control block for empty sets' begin/end. I assume it's so that dereferencing them will fail? However, isn't this changing the semantics in the sense that these iterators won't be tied to the specific set anymore? So would this, e.g., succeed for two empty sets: `empty1.begin() == empty2.begin()`? (not sure how std::iterators compare when pointing to different sets).

Reply:
Yes, this is done to make deref of `begin()` of an empty set throw. As for the comparison, with this implementation we cannot see that the iterators point to different sets, and defer to the comparison of the underlying stdlib iterators. These _should_ compare not equal for different sets, but I think this is implementation-defined. As an alternative, one could perform a `size` check to make sure the set is not empty when dereferencing an iterator, or introduce some sentinel iterator type for empty sets, but I feel this implementation here is good enough. Let's definitely add a test.